compiler/modules/front/resolveTypes.bal

```diff
@@ -28,6 +28,52 @@ function resolveTypes(ModuleSymbols mod) returns ResolveTypeError? {
     foreach var defn in mod.defns {
         check resolveDefn(mod, defn);
     }
+    check finishEmptinessChecks(mod);
+}
+
+function finishEmptinessChecks(ModuleSymbols mod) returns ResolveTypeError? {
+    if !mod.tc.env.isReady() {
+        // This should never happen
+        panic err:impossible("type environment is not ready");
+    }
+    map<DeferredEmptinessCheck[]> deferredEmptinessChecks = mod.deferredEmptinessChecks;
+    foreach var [name, emptinessChecks] in deferredEmptinessChecks.entries() {
+        foreach var { semType, modDefn } in emptinessChecks {
+            // XXX When we can give multiple errors we should check all the deferred emptiness checks
+            check nonEmptyTypeDeferred(mod, semType, modDefn);
+            if !mod.deferredEmptinessChecks.hasKey(name) {
+                break;
+            }
+        }
+        // This should never happen
+        if mod.deferredEmptinessChecks.hasKey(name) {
+            DeferredEmptinessCheck[] remainingChecks = deferredEmptinessChecks.get(name);
+            var { modDefn } = remainingChecks[0];
+            d:Location loc = s:locationInDefn(modDefn, { startPos: modDefn.startPos, endPos: modDefn.endPos });
+            return err:semantic("non recursive empty type", loc);
+        }
+    }
+}
+
+function nonEmptyTypeDeferred(ModuleSymbols mod, t:SemType semType, s:ModuleLevelDefn modDefn) returns ResolveTypeError? {
+    if t:isEmpty(mod.tc, semType) {
+        s:TypeDesc td = modDefn is s:FunctionDefn ? modDefn.typeDesc : <s:TypeDesc>modDefn.td;
+        d:Location loc = s:locationInDefn(modDefn, { startPos: td.startPos, endPos: td.endPos });
+        if td is s:BinaryTypeDesc && td.op is "&" {
+            _ = mod.deferredEmptinessChecks.removeIfHasKey(modDefn.name); // Currently this has no effect
+            return err:semantic("intersection must not be empty", loc);
+        }
+        // We are only defering intersections(already handled), lists and mappings
+        t:ComplexSemType t = <t:ComplexSemType>semType;
```

Comment:
This module shouldn't know about the internals of how the types module deals with atomic types.

I think it would be better to the types module to have a function that takes a SemType and tests whether it is atomic and recursive.

Reply:
fixed